wallet/types/src/keys.rs

```diff
@@ -13,9 +13,94 @@
 use crypto::key::extended::{ExtendedPrivateKey, ExtendedPublicKey};
+use crypto::key::hdkd::child_number::ChildNumber;
+use crypto::key::hdkd::u31::U31;
 use serialization::{Decode, Encode};
 
+/// The usage purpose of a key i.e. if it is for receiving funds or for change
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+#[allow(clippy::unnecessary_cast)]
+pub enum KeyPurpose {
+    /// This is for addresses created for receiving funds that are given to the user
+    ReceiveFunds = 0,
+    /// This is for the internal usage of the wallet when creating change output for a transaction
+    Change = 1,
+}
+
+impl KeyPurpose {
+    /// All purposes
+    pub const ALL: [KeyPurpose; 2] = [ReceiveFunds, Change];
+    /// The index for receiving funds
+    const RECEIVE_FUNDS_INDEX: ChildNumber = ChildNumber::from_normal(U31::from_u32_with_msb(0).0);
+    /// The index for change addresses/keys
+    const CHANGE_INDEX: ChildNumber = ChildNumber::from_normal(U31::from_u32_with_msb(1).0);
+
+    pub fn get_deterministic_index(&self) -> ChildNumber {
+        match self {
+            ReceiveFunds => Self::RECEIVE_FUNDS_INDEX,
+            Change => Self::CHANGE_INDEX,
+        }
+    }
+}
+
+impl TryFrom<ChildNumber> for KeyPurpose {
+    type Error = ChildNumber;
+
+    fn try_from(num: ChildNumber) -> Result<Self, Self::Error> {
+        match num.get_index() {
+            0 => Ok(ReceiveFunds),
+            1 => Ok(Change),
+            _ => Err(num),
+        }
+    }
+}
+
+// TODO store separately
+// receiving_state: KeychainUsageState,
+// change_state: KeychainUsageState,
+
+/// Struct that holds information for account addresses
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+pub struct KeychainUsageState {
+    /// Last used address index. An address might not be used so the corresponding entry would be
+    /// None, otherwise it would be that last used ChildNumber
+    last_used: Option<ChildNumber>,
+    /// Last issued address to the user. Those addresses can be issued until the
+    /// last used index + lookahead size.
+    last_issued: Option<ChildNumber>,
+}
+
+impl KeychainUsageState {
+    pub fn new(last_used: Option<ChildNumber>, last_issued: Option<ChildNumber>) -> Self {
+        Self {
+            last_used,
+            last_issued,
+        }
+    }
+
+    /// Get the last index used in the blockchain
+    pub fn get_last_used(&self) -> Option<ChildNumber> {
+        self.last_used
+    }
+
+    /// Set the last index used in the blockchain
+    pub fn set_last_used(&mut self, new_last_used: Option<ChildNumber>) {
+        self.last_used = new_last_used;
+    }
+
+    /// Get the last index issued to the user
+    pub fn get_last_issued(&self) -> Option<ChildNumber> {
+        self.last_issued
+    }
+
+    /// Set the last index issued to the user
+    pub fn set_last_issued(&mut self, new_last_issued: Option<ChildNumber>) {
```

Comment:
If the only way we use this is by incrementing the value, why does this struct expose a free functionality to do anything with the number? Why not just increment it here, which guarantees the behavior to external users of this module?